### PR TITLE
[WIP] Add TFJob/PyTorchJob support

### LINKED ISSUE #39: [feature] Integrate with tf-operator, pytorch-operator
1. take parameters that control distributed training (e.g. number of replicas)
2. create the appropriate CRD as opposed to K8s job controller

https://kubeflow.slack.com/archives/C9ZLKR73L/p1523553472000001

### PR DESCRIPTION
**Pull request by gaocegege**
Close #39 

Signed-off-by: Ce Gao <gaoce@caicloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/227)

### CODE REVIEW

**Comment by gaocegege** [comment]
/assign @YujiOshima 

I am working on the support of TFJob. I added TFJob as a new kind of worker raw template, and did corresponding changes in metrics collector and controller, could you please reiview if the way works for you?

/cc @ddysher @jlewi @richardsliu

**Comment by YujiOshima** [comment]
@gaocegege Great, Thanks!  This is the very thing I wanted.

**Comment by gaocegege** [comment]
Cool, then I will keep going.

File: pkg/controller/studyjobcontroller/const.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package studyjobcontroller
+
+const (
+	workerKindJob   = "Job"
```
Comment (johnugeorge) [inline pkg/controller/studyjobcontroller/const.go]: how about WorkerDefaultKind, WorkerTFKind to give more clarity?

Comment (gaocegege) [reply pkg/controller/studyjobcontroller/const.go]: SGTM

**Comment by richardsliu** [comment]
@gaocegege What is the status of this PR?

I have some free cycles in the next few weeks, I can take this up and complete the implementation.

**Comment by johnugeorge** [comment]
After Kubecon(next week), I can too help to fix the pending issues for coming release.

**Comment by gaocegege** [comment]
@richardsliu @johnugeorge 

Thanks for your time. I am working on my thesis these days. If you could take this up, I'd appreciate it.

File: examples/random-tfjob-example.yaml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      parametertype: categorical
+      feasible:
+        list:
+        - sgd
+        - adam
+        - ftrl
+  workerSpec:
+    goTemplate:
+        rawTemplate: |-
+          apiVersion: "kubeflow.org/v1alpha1"
```
Comment (richardsliu) [inline examples/random-tfjob-example.yaml]: This should be using the v1beta1 API now. V1apha1 is deprecated.

**Comment by gaocegege** [comment]
@richardsliu Feel free to clone the PR 😄

File: examples/random-tfjob-example.yaml [hunk 3]
```diff
@@ -0,0 +1,72 @@
+apiVersion: "kubeflow.org/v1alpha1"
+kind: StudyJob
+metadata:
+  namespace: katib
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: random-example
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply examples/random-tfjob-example.yaml]: SGTM

File: examples/random-tfjob-example.yaml (hunk 2 above)
Comment (gaocegege) [reply examples/random-tfjob-example.yaml]: SGTM

**Comment by gaocegege** [comment]
I am going to close the PR, feel free to fork it. And I am free to review and give you suggestions about the implementation.

Again, thank @richardsliu !

**Pull request closed without merging**